crates/core/tedge_agent/src/http_server/entity_store.rs

```diff
@@ -170,13 +181,113 @@ async fn register_entity(
+            .await
+            .map(|res| res.into_response()),
+        Channel::EntityTwinData { fragment_key } => {
+            get_entity_twin_fragment(state, topic_id, fragment_key.to_string())
+                .await
+                .map(|res| res.into_response())
+        }
+        _ => Err(Error::PathNotFound(path)),
+    }
+}
+
+async fn put_resource(
+    State(state): State<AgentState>,
+    Path(path): Path<String>,
+    Json(value): Json<Value>,
+) -> impl IntoResponse {
+    let (topic_id, channel) = parse_path(&path)?;
+    match channel {
+        Channel::EntityTwinData { fragment_key } => {
+            Ok(upsert_entity_twin_fragment(state, topic_id, fragment_key.to_string(), value).await)
+        }
+        _ => Err(Error::PathNotFound(path)),
+    }
+}
+
+async fn patch_resource(
     State(state): State<AgentState>,
     Path(path): Path<String>,
     Json(twin_fragments): Json<Map<String, Value>>,
 ) -> impl IntoResponse {
-    let topic_id = EntityTopicId::from_str(&path)?;
-    let twin_data = EntityTwinData::try_new(topic_id, twin_fragments)?;
+    let (topic_id, channel) = parse_path(&path)?;
+    match channel {
+        Channel::EntityMetadata => Ok(patch_entity(state, topic_id, twin_fragments).await),
+        _ => Err(Error::PathNotFound(path)),
+    }
+}
+
+async fn delete_resource(
+    State(state): State<AgentState>,
+    Path(path): Path<String>,
+) -> Result<Response, Error> {
+    let (topic_id, channel) = parse_path(&path)?;
+    match channel {
+        Channel::EntityMetadata => deregister_entity(state, topic_id).await,
+        Channel::EntityTwinData { fragment_key } => {
+            delete_entity_twin_fragment(state, topic_id, fragment_key.to_string()).await
+        }
+        _ => Err(Error::PathNotFound(path)),
+    }
+}
+
+fn parse_path(path: &str) -> Result<(EntityTopicId, Channel), Error> {
+    let segments = path.split('/').collect::<Vec<&str>>();
+    match segments.as_slice() {
+        [seg1, seg2] => {
+            let topic_id = topic_id_from_path_segments(seg1, Some(seg2), None, None)?;
+            Ok((topic_id, Channel::EntityMetadata))
+        }
+        [seg1, seg2, seg3] => {
+            let topic_id = topic_id_from_path_segments(seg1, Some(seg2), Some(seg3), None)?;
+            Ok((topic_id, Channel::EntityMetadata))
+        }
+        [seg1, seg2, seg3, seg4] => {
+            let topic_id = topic_id_from_path_segments(seg1, Some(seg2), Some(seg3), Some(seg4))?;
+            Ok((topic_id, Channel::EntityMetadata))
+        }
+        [seg1, seg2, seg3, seg4, "twin", fragment_key] => {
+            let topic_id = topic_id_from_path_segments(seg1, Some(seg2), Some(seg3), Some(seg4))?;
+            Ok((
+                topic_id,
+                Channel::EntityTwinData {
+                    fragment_key: fragment_key.to_string(),
+                },
+            ))
+        }
+        _ => Err(Error::PathNotFound(path.to_string())),
```

Comment:
The error cases must give move insight to the user.

Currently, the followings lead to the same error message telling the path is not found.

* Incorrect channel: `device/main///cmd/maintenance_window`
* Missing fragment key: `device/main///twin`
* Incorrect fragment key: `device/main///twin/maintenance/window not`

Reply:
I was just trying to mimic the simple `Not Found` message emitted by axum when a non-existent resource path is queried. Will fix that by adding more context.

Reply:
We should also prevent empty fragment keys as well as the following leads to an empty key value.

```sh
$ tedge http put '/tedge/entity-store/v1/entities/device/main///twin/' '{"hello":"foo"}'
{"@topic-id":"device/main//","@type":"device","":{"hello":"foo"},"hello":"foo2","hello2":"foo2"}
```

Reply:
> Resolved by f837935c

Now, working as expected